storage/csv/datasource.go

```diff
@@ -145,66 +320,86 @@ func (rs *RecordStream) initializeColumnsWithHeaderRow() error {
+	aliasedRecord := make(map[octosql.VariableName]octosql.Value)
+	for i, v := range line {
+		aliasedRecord[rs.aliasedFields[i]] = execution.ParseType(v)
+	}
+
+	return aliasedRecord, nil
 }
 
-func (rs *RecordStream) Next(ctx context.Context) (*execution.Record, error) {
-	if rs.isDone {
-		return nil, execution.ErrEndOfStream
+func parseDataTypes(row []string) []octosql.Value {
+	resultRow := make([]octosql.Value, len(row))
+	for i, v := range row {
+		resultRow[i] = execution.ParseType(v)
 	}
 
-	if rs.first {
-		rs.first = false
+	return resultRow
+}
 
-		if rs.hasColumnHeader {
-			err := rs.initializeColumnsWithHeaderRow()
-			if err != nil {
-				return nil, errors.Wrap(err, "couldn't initialize columns for record stream")
-			}
+var offsetPrefix = []byte("csv_offset")
 
-			return rs.Next(ctx)
-		} else {
-			record, err := rs.initializeColumnsWithoutHeaderRow()
-			if err != nil {
-				return nil, errors.Wrap(err, "couldn't initialize columns for record stream")
-			}
+func (rs *RecordStream) loadOffset(tx storage.StateTransaction) error {
+	offsetState := storage.NewValueState(tx.WithPrefix(offsetPrefix))
 
-			return record, nil
-		}
+	var offset octosql.Value
+	err := offsetState.Get(&offset)
+	if err == storage.ErrNotFound {
+		offset = octosql.MakeInt(0)
+	} else if err != nil {
+		return errors.Wrap(err, "couldn't load csv offset from state storage")
 	}
 
-	line, err := rs.r.Read()
-	if err == io.EOF {
-		rs.isDone = true
-		rs.file.Close()
-		return nil, execution.ErrEndOfStream
-	}
+	rs.offset = offset.AsInt()
+
+	return nil
+}
 
+func (rs *RecordStream) saveOffset(tx storage.StateTransaction, curBatchSize int) error {
+	offsetState := storage.NewValueState(tx.WithPrefix(offsetPrefix))
+
+	rs.offset = rs.offset + curBatchSize
+
+	offset := octosql.MakeInt(rs.offset)
+	err := offsetState.Set(&offset)
 	if err != nil {
-		return nil, errors.Wrap(err, "couldn't read record")
+		return errors.Wrap(err, "couldn't save csv offset to state storage")
 	}
 
-	aliasedRecord := make(map[octosql.VariableName]octosql.Value)
-	for i, v := range line {
-		aliasedRecord[rs.aliasedFields[i]] = execution.ParseType(v)
+	return nil
+}
+
+func (rs *RecordStream) Next(ctx context.Context) (*execution.Record, error) {
+	tx := storage.GetStateTransactionFromContext(ctx).WithPrefix(rs.streamID.AsPrefix())
+	outputQueue := execution.NewOutputQueue(tx.WithPrefix(outputQueuePrefix))
+
+	var queueElement QueueElement
+	err := outputQueue.Pop(ctx, &queueElement)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't pop queue element")
 	}
 
-	curOffset := rs.offset
-	rs.offset++
-	return execution.NewRecord(
-		rs.aliasedFields,
-		aliasedRecord,
-		execution.WithID(execution.NewRecordIDFromStreamIDWithOffset(rs.streamID, curOffset)),
-	), nil
+	switch queueElement := queueElement.Type.(type) {
+	case *QueueElement_Record:
+		return queueElement.Record, nil
+	case *QueueElement_Error:
+		if queueElement.Error == execution.ErrEndOfStream.Error() {
```

Comment:
Don't compare strings. Add a message type EndOfStream and send it. Same as in join or ProcessByKey.